File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:        
        
        if not subRoot:
            return True
        if not root:
            return False

        if self.equalTree(root, subRoot):
            return True
        left = self.isSubtree(root.left, subRoot) 
        right = self.isSubtree(root.right, subRoot)

        return left or right

    def equalTree(self, root, subRoot) -> bool:
            if not root and not subRoot:
                return True

            if root and subRoot and root.val == subRoot.val:
                left = self.equalTree(root.left, subRoot.left)
                right = self.equalTree(root.right, subRoot.right)
            
                return left and right
            return False

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0 || inorder.length == 0){
            return null;
        }

        int root = preorder[0];
        int mid = -1;

        for (int i = 0; i < inorder.length; i++){
            if (root == inorder[i]){
                mid = i;
                break;
            }
        }

        int[] leftPreorder = Arrays.copyOfRange(preorder, 1, mid+1);
        int[] leftInorder = Arrays.copyOfRange(inorder, 0, mid);
        TreeNode leftNode = buildTree(leftPreorder, leftInorder);

        int[] rightPreorder = Arrays.copyOfRange(preorder, mid+1, preorder.length);
        int[] rightInorder = Arrays.copyOfRange(inorder, mid+1, inorder.length);
        TreeNode rightNode = buildTree(rightPreorder, rightInorder);

        TreeNode rootNode = new TreeNode(root);
        rootNode.left = leftNode;
        rootNode.right = rightNode;

        return rootNode;

    }
}
